### ISSUE TITLE
log the OpenGL errors

### ISSUE THREAD

**Issue by LovelySanta** [TheCherno/Hazel#75]
As mensioned in discord, currently we're not logging any openGL errors that could possible occur. To do this, I would recommend creating the same macro as Cherno did in [his OpenGL series](https://www.youtube.com/watch?v=FBbPWSOQ0-w&list=PLlrATfBNZ98foTJPJ_Ev03o2oq3-GGOS2&index=10).

### Step 1: Creating the macro
In [src/Platform/OpenGL/](https://github.com/TheCherno/Hazel/tree/master/Hazel/src/Platform/OpenGL) create a new file `OpenGLMacro.h` where the macro is defined. We do it in the OpenGL folder as it is OpenGL specific:
```cpp
// OpenGLMacro.h
#pragma once
#include "Engine/Core.h"
#include "Engine/Log.h"

#include "glad/glad.h"

inline static void GLClearError() { while (glGetError() != GL_NO_ERROR); }

static bool GLLogCall(const char* function, const char* file, const int line)
{
    while (GLenum error = glGetError())
    {
        ENGINE_CORE_ERROR("[OpenGL] Error ({0}):", error);
        ENGINE_CORE_ERROR("[OpenGL]     in function {0}", function);
        ENGINE_CORE_ERROR("[OpenGL]     at {0} ({1})", file, line);
        return false;
    }
    return true;
}

#if defined ENGINE_DEBUG
    #define GLCall(x) GLClearError();\
                      x;\
                      ENGINE_CORE_ASSERT(GLLogCall(#x, __FILE__, __LINE__), "OpenGL Error!")
#else
    #define GLCall(x) x
#endif
```

As described in Cherno's video, the macro wraps two functions around the actual call:

- `GLClearError()` before the function call to clear all previous errors
- `GLLogCall(...)` after the function to log any errors.

_**NOTE:**_ `ENGINE_CORE_ASSERT` currently does not call the function outside the debug configuration, which is here not an issue as the GLCall macro is only defined inside the debug configuration.

When `ENGINE_DEBUG` is not defined, GLCall will just call the function without any other fuctions wrapped around it.

### Step 2: Using the macro
In every OpenGl cpp file (inside [src/Platform/OpenGL/](https://github.com/TheCherno/Hazel/tree/master/Hazel/src/Platform/OpenGL)) we include the OpenGLMacro.h and add `GLCall()` around each raw OpenGL call.

For example in [src/Platform/OpenGL/OpenGLBuffer.cpp line 1-15:](https://github.com/TheCherno/Hazel/blob/master/Hazel/src/Platform/OpenGL/OpenGLBuffer.cpp#L1-L15):
```cpp
#include "hzpch.h"
#include "OpenGLBuffer.h"

#include <glad/glad.h>

namespace Hazel {

	/////////////////////////////////////////////////////////////////////////////
	// VertexBuffer /////////////////////////////////////////////////////////////
	/////////////////////////////////////////////////////////////////////////////

	OpenGLVertexBuffer::OpenGLVertexBuffer(float* vertices, uint32_t size)
	{
		glCreateBuffers(1, &m_RendererID);
```
becomes
```cpp
#include "hzpch.h"
#include "OpenGLBuffer.h"
#include "OpenGLMacro.h"

#include <glad/glad.h>

namespace Hazel {

	/////////////////////////////////////////////////////////////////////////////
	// VertexBuffer /////////////////////////////////////////////////////////////
	/////////////////////////////////////////////////////////////////////////////

	OpenGLVertexBuffer::OpenGLVertexBuffer(float* vertices, uint32_t size)
	{
		GLCall(glCreateBuffers(1, &m_RendererID));
```

_**NOTE:**_ This could also be added in the current shader class implementation, but I would leave that out until the Shader class is abstracted to an OpenGLShader

**Comment by TheCherno** [comment]
This isn't necessary, as we can use `glDebugMessageCallback` instead of polling for errors after every single GL command. Minimum spec for this is 4.3, but as it stands I'm not sure I have any interest in supporting OpenGL versions below that, so it should be fine. See https://www.khronos.org/opengl/wiki/Debug_Output for more info.

**Comment by LovelySanta** [comment]
Since recent we have an init function:
https://github.com/TheCherno/Hazel/blob/e8b2ad932372b0121349f34df14e743bb95773fc/Hazel/src/Platform/OpenGL/OpenGLRendererAPI.cpp#L8-L12

Here we should register the `glDebugMessageCallback` as cherno wants.

**Comment by ilikenoodles2** [comment]
If this is still needed, I could implement a pull request because it seems one hasn't been created yet.

Also, with glDebugMessageCallback, there needs to be a debug OpenGL context.  Question is, should a debug context be used with only Debug builds, or should a debug context be in both Debug and Release builds?

**Comment by LovelySanta** [comment]
You don't need a context for it?

Create a function `static void OpenGLMessage(...)` in the OpenGLRendererAPI and at the beginning of the Init function you can set the callback & enable the logging.

In Core.h we have a part
```
#ifdef HZ_DEBUG
	#define HZ_ENABLE_ASSERTS
#endif
```

You could add `#define HZ_ENABLE_RENDERER_LOG` and use that to wrap the statements in the OpenGLRendererAPI so we can control the debugging in core.h in the future.

Extra:
Make a macro for `Hazel::Log::Init()` call it `HZ_LOG_INIT()` (in Log.h)
Then in the same place where you added `#define HZ_ENABLE_RENDERER_LOG` also define `#define HZ_ENABLE_ENGINE_LOG` & `#define HZ_ENABLE_CLIENT_LOG` and use that to wrap around all the macros in Log.h so we can disable all logging at once.

That should work nicely?

**Comment by WhoseTheNerd** [comment]
Using debug output requires you to create special context, context with debug flag enabled. Then register the callback.

**Comment by LovelySanta** [comment]
> Using debug output requires you to create special context, context with debug flag enabled. Then register the callback.

Which is implemented in #114: 
https://github.com/TheCherno/Hazel/blob/784f46ff9dd93e5adc6aacb3bf757054225a8a33/Hazel/src/Platform/OpenGL/OpenGLRendererAPI.cpp#L30-L36
https://github.com/TheCherno/Hazel/blob/784f46ff9dd93e5adc6aacb3bf757054225a8a33/Hazel/src/Platform/OpenGL/OpenGLRendererAPI.cpp#L8-L26
